presto-main/src/main/java/com/facebook/presto/sql/planner/CachingPlanAnalyticsInfoProvider.java

```diff
@@ -0,0 +1,89 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.common.plan.PlanCanonicalizationStrategy;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.statistics.PlanStatistics;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.Weigher;
+import io.airlift.units.DataSize;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.common.type.encoding.StringUtils.UTF_8;
+import static com.google.common.hash.Hashing.sha256;
+
+public class CachingPlanAnalyticsInfoProvider
+        implements PlanCanonicalInfoProvider
+{
+    private static final DataSize CACHE_SIZE_BYTES = new DataSize(2, DataSize.Unit.MEGABYTE);
+    private final Cache<PlanNode, String> cache =
+            CacheBuilder.newBuilder()
+                    .maximumWeight(CACHE_SIZE_BYTES.toBytes())
+                    .weigher((Weigher<PlanNode, String>) (key, value) -> value.length() + 20)
+                    .expireAfterWrite(5, TimeUnit.MINUTES)
+                    .build();
+
+    private final ObjectMapper objectMapper;
+    private final Metadata metadata;
+
+    public CachingPlanAnalyticsInfoProvider(ObjectMapper objectMapper, Metadata metadata)
+    {
+        this.objectMapper = objectMapper;
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Optional<String> hash(Session session, PlanNode planNode, PlanCanonicalizationStrategy strategy)
+    {
+        String result = cache.getIfPresent(planNode);
+        if (result != null) {
+            return Optional.of(result);
+        }
+        CanonicalPlanGenerator.Context context = new CanonicalPlanGenerator.Context();
+        planNode.accept(
+                new CanonicalPlanGenerator(PlanCanonicalizationStrategy.EXACT, objectMapper, session),
+                context);
+        context
+                .getCanonicalPlans()
+                .forEach(
+                        (p, canonicalPlan) -> {
+                            cache.put(p, hashCanonicalPlan(canonicalPlan, objectMapper));
+                        });
+        return Optional.ofNullable(cache.getIfPresent(planNode));
+    }
+
+    @Override
+    public Optional<List<PlanStatistics>> getInputTableStatistics(Session session, PlanNode planNode)
+    {
+        return Optional.empty();
+    }
+
+    public void invalidateCache()
+    {
+        cache.invalidateAll();
```

Comment:
We cannot invalidate the whole cache, because this cache can be used by multiple queries. 

You can compute all this only once per query, doing all the work i historybasedstatisticstracker.

Reply:
Ahh damn true, nice catch!
 **getPlanAnalyticsQueryStats(QueryInfo queryInfo)** in historybasedstatisticstracker uses QueryInfo which does not have the plan. So either we add the plan there or we can do a per query map as a cache.
in [getPlanAnalyticsCanonicalInfo](https://github.com/prestodb/presto/blob/20d116133fd9bce796da1e9d948ce008e830b967/presto-main/src/main/java/com/facebook/presto/sql/planner/PlanNodeCanonicalInfo.java#L119).

like adding  `hash(Session session, PlanNode planNode, PlanCanonicalizationStrategy strategy, Map<PlanNode, String> perQueryPlanNodeCacheMap)` to the PlanCanonicalInfoProvider interface.

Reply:
I will try and check the most efficient approach

Reply:
In [commit](https://github.com/prestodb/presto/pull/19557/commits/1af8c22a58fd12851b151d42d50196f07daeb6c4), I made the planAnalyticsInfoProvider accept a context which basically is a per Query cache. 
Maybe in future the CachingPlanCanonicalInfoProvider can be refactored to hash based on a per query cache along with the global hash according to different requirements.